chia/full_node/subscriptions.py

```diff
@@ -1,144 +1,179 @@
+
+            for item in items:
+                row = conn.execute(
+                    f"SELECT COUNT(*) FROM {table_name} WHERE peer_id=? AND {item_name}=?",
+                    (peer_id, item),
+                ).fetchone()
+
+                assert row is not None
+
+                if int(row[0]) > 0:
+                    continue
+
+                conn.execute(f"INSERT INTO {table_name} (peer_id, {item_name}) VALUES (?, ?)", (peer_id, item))
+                inserted.add(item)
+                subscriptions_left -= 1
+
+                if subscriptions_left == 0:
+                    log.info(
+                        "Peer %s reached the subscription limit. Not all coin states will be reported.",
+                        peer_id,
+                    )
+                    break
+
+            return inserted
+
+    def add_puzzle_subscriptions(self, peer_id: bytes32, puzzle_hashes: List[bytes32], max_items: int) -> Set[bytes32]:
+        """
+        Returns the items that were actually subscribed to, which is fewer in these cases:
+        * There are duplicate items.
+        * Some items are already subscribed to.
+        * The `max_items` limit is exceeded.
+        """
+        return self._add_subscriptions(
+            peer_id=peer_id,
+            table_name="puzzle_subscriptions",
+            item_name="puzzle_hash",
+            items=puzzle_hashes,
+            max_items=max_items,
+        )
+
+    def add_coin_subscriptions(self, peer_id: bytes32, coin_ids: List[bytes32], max_items: int) -> Set[bytes32]:
+        """
+        Returns the items that were actually subscribed to, which is fewer in these cases:
+        * There are duplicate items.
+        * Some items are already subscribed to.
+        * The `max_items` limit is exceeded.
+        """
+        return self._add_subscriptions(
+            peer_id=peer_id,
+            table_name="coin_subscriptions",
+            item_name="coin_id",
+            items=coin_ids,
+            max_items=max_items,
+        )
+
+    def peer_subscription_count(self, peer_id: bytes32) -> int:
+        with self._db_conn as conn:
+            row = conn.execute("SELECT COUNT(*) FROM puzzle_subscriptions WHERE peer_id=?", (peer_id,)).fetchone()
+            assert row is not None
+            puzzle_subscription_count = int(row[0])
+
+            row = conn.execute("SELECT COUNT(*) FROM coin_subscriptions WHERE peer_id=?", (peer_id,)).fetchone()
+            assert row is not None
+            coin_subscription_count = int(row[0])
+
+        return puzzle_subscription_count + coin_subscription_count
+
+    def total_puzzle_subscriptions(self) -> int:
+        with self._db_conn as conn:
+            row = conn.execute("SELECT COUNT(*) FROM puzzle_subscriptions").fetchone()
+            assert row is not None
+            puzzle_subscription_count = int(row[0])
+
+        return puzzle_subscription_count
+
+    def total_coin_subscriptions(self) -> int:
```

Comment:
I believe `peer_subscription_count()`, `total_puzzle_subscriptions()` and `total_coin_subscriptions()` are missing unit tests